Dynamic lora load/unload sidecar (#31)
kubernetes-sigs/llm-instance-gateway: tools/dynamic-lora-sidecar/README.md

# Dynamic LORA Adapter Sidecar for vLLM

> This directory contains a configmap containing lora adapters configurations and script for a sidecar container to dynamically manage LORA adapters for a vLLM server running in the same Kubernetes pod by reconciling it with a configmap containing lora adapters. 

Reviewer: ```suggestion
This is a sidecar-based tool to help rolling out new LoRA adapters to a set of running vLLM model servers. The user deploys the sidecar with a vLLM server, and using a ConfigMap, the user can express their intent as to which LoRA adapters they want to have the running vLLM servers to be configure with. The sidecar watches the ConfigMap and sends load/unload requests to the vLLM container to actuate on the user intent. 
```

[deployment]: deployment.yaml it uses [sidecar](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/)(`initContainer` with `restartPolicy` set to `always`) which is beta feature enabled by default since k8s version 1.29. They need to be enabled in 1.28 and prior to 1.28 sidecar are not officially supported.

## Configuration Fields
- `vLLMLoRAConfig`[**required**]  base key 
- `host` [*optional*]Model server's host. defaults to localhost
- `port` [*optional*] Model server's port. defaults to 8000
- `name`[*optional*] Name of this config
- `ensureExist`[*optional*] List of models to ensure existence on specified model server.
    -  `models`[**required**] [list]
        - `base-model`[*optional*] Base model for lora adapter
        - `id`[**required**] unique id of lora adapter
        - `source`[**required**] path (remote or local) to lora adapter
- `ensureNotExist` [*optional*]
    - `models`[**required**] [list]
        - `id`[**required**] unique id of lora adapter
        -  `source`[**required**] path (remote or local) to lora adapter
        - `base-model`[*optional*] Base model for lora adapter

## Screenshots & Testing

> I tested the sidecar in my cluster with deployment and configmap specified in this repo. Here are the screen grabs of the logs from the sidecar and vllm server. I used the specified configmap, verified that the adapters were loaded by querying `v1/models` and looking at vllm logs. I changed the configmap and validated the same on vllm server. Note: There is slight lag between updates. 

Reviewer: ```suggestion
The sidecar was tested with the Deployment and ConfigMap specified in this repo. Here are screen grabs of the logs from the sidecar and vllm server. One can verify that the adapters were loaded by querying `v1/models` and looking at vllm logs.
```
